test/e2e/instrumentation/monitoring/custom_metrics-stackdriver.go

```diff
@@ -0,0 +1,200 @@
+	var customMetricsClient customclient.CustomMetricsClient
+	var discoveryClient *discovery.DiscoveryClient
+
+	It("should run Custom Metrics - Stackdriver Adapter [Feature:StackdriverMonitoring]", func() {
+		kubeClient = f.ClientSet
+		kubeAggrClient = f.AggregatorClient
+		config, err := framework.LoadConfig()
+		if err != nil {
+			framework.Failf("Failed to load config: %s", err)
+		}
+		customMetricsClient = customclient.NewForConfigOrDie(config)
+		discoveryClient = discovery.NewDiscoveryClientForConfigOrDie(config)
+		testAdapter(f, kubeClient, customMetricsClient, discoveryClient)
+	})
+})
+
+func testAdapter(f *framework.Framework, kubeClient clientset.Interface, customMetricsClient customclient.CustomMetricsClient, discoveryClient *discovery.DiscoveryClient) {
+	projectId := framework.TestContext.CloudConfig.ProjectID
+
+	ctx := context.Background()
+	client, err := google.DefaultClient(ctx, gcm.CloudPlatformScope)
+
+	// Hack for running tests locally
+	// If this is your use case, create application default credentials:
+	// $ gcloud auth application-default login
+	// and uncomment following lines (comment out the two lines above):
+	/*
+		ts, err := google.DefaultTokenSource(oauth2.NoContext)
+		framework.Logf("Couldn't get application default credentials, %v", err)
+		if err != nil {
+			framework.Failf("Error accessing application default credentials, %v", err)
+		}
+		client := oauth2.NewClient(oauth2.NoContext, ts)
+	*/
+
+	gcmService, err := gcm.New(client)
+	if err != nil {
+		framework.Failf("Failed to create gcm service, %v", err)
+	}
+
+	framework.ExpectNoError(err)
+
+	// Set up a cluster: create a custom metric and set up k8s-sd adapter
+	err = createDescriptors(gcmService, projectId)
+	if err != nil {
+		framework.Failf("Failed to create metric descriptor: %s", err)
+	}
+	defer cleanupDescriptors(gcmService, projectId)
+
+	err = createAdapter()
+	if err != nil {
+		framework.Failf("Failed to set up: %s", err)
+	}
+	defer cleanupAdapter()
+
+	// Run application that exports the metric
+	err = createSDExporterPod(kubeClient)
+	if err != nil {
+		framework.Failf("Failed to create sd-exporter pod: %s", err)
+	}
+	defer cleanupSDExporterPod(kubeClient)
+
+	// Wait a short amount of time to create a pod and export some metrics
+	time.Sleep(60 * time.Second)
+
+	// Verify responses from Custom Metrics API
+	resources, err := discoveryClient.ServerResourcesForGroupVersion("custom.metrics.k8s.io/v1beta1")
+	if err != nil {
+		framework.Failf("Failed to retrieve a list of supported metrics: %s", err)
+	}
+	for _, resource := range resources.APIResources {
+		if resource.Name != "pods/"+CustomMetricName && resource.Name != "pods/"+UnusedMetricName {
+			framework.Failf("Unexpected metric %s. Only metric %s should be supported", resource.Name, CustomMetricName)
+		}
+	}
+	value, err := customMetricsClient.NamespacedMetrics("default").GetForObject(schema.GroupKind{Group: "", Kind: "Pod"}, "sd-exporter-1", CustomMetricName)
```

Comment:
Consider splitting this test into smaller methods, e.g. verifyPodMetrics, verifyPodSelectorMetrics

Reply:
What do you think would it improve? There is no repeated code that can be extracted. Do you think this would be more readable if each case would have it's own verify function? I prefer to leave it as it is, but I can change it if you think it would affect readability in a significant way.